Consider the dataset in Cases/Bankruptcy. Y=D, X:R1,,,R24 Run the MLP with PyTorch Code with Layers as ( 24,  15,  10, 5, 1)   and  ( 24, 13, 7, 1)

In [50]:
import torch
import pandas as pd
import numpy as np
import torch.nn as nn
from sklearn.model_selection import train_test_split 
from sklearn.metrics import accuracy_score, log_loss,r2_score
from sklearn.preprocessing import MinMaxScaler

In [2]:
bank=pd.read_csv("Bankruptcy.csv",index_col=0)

In [3]:
bank

D  YR    R1    R2    R3    R4    R5    R6    R7     R8  ...   R15   R16  \
NO                                                           ...               
1    0  78  0.23  0.08  0.02  0.03  0.46  0.12  0.19  10.36  ...  0.05  0.57   
2    0  77  0.19  0.07  0.09  0.12  0.02  0.02  0.03   3.13  ...  0.09  0.12   
3    0  72  0.07  0.02  0.03  0.05  0.06  0.10  0.14   2.41  ... -0.03  0.02   
4    0  80  0.07  0.03  0.04  0.04  0.04  0.06  0.06   5.55  ... -0.02  0.01   
5    0  81  0.09  0.02  0.03  0.04  0.06  0.08  0.11   2.85  ...  0.02  0.07   
..  ..  ..   ...   ...   ...   ...   ...   ...   ...    ...  ...   ...   ...   
128  1  77  0.27  0.03  0.07  0.10  0.09  0.18  0.26   7.58  ...  0.11  0.06   
129  1  77  0.32  0.03  0.03  0.09  0.05  0.06  0.16   3.24  ...  0.17  0.07   
130  1  78  0.08  0.01  0.02  0.05  0.04  0.07  0.15   3.82  ...  0.19  0.07   
131  1  78  0.14  0.01  0.05  0.07  0.02  0.09  0.14   8.18  ...  0.07  0.02   
132  1  78  0.26  0.07  0.12  0.17  0.00  0.01  0.01   6.13  ...  0.05  0.04   

      R17   R18     R19   R20   R21   R22   R23   R24  
NO                                                     
1    0.15  0.23    3.56  0.26  1.55  0.43  0.11  0.17  
2    0.16  0.22    3.78  1.29  1.40  0.06  0.07  0.10  
3    0.02  0.04   13.29  1.61  1.43  0.03  0.05  0.07  
4    0.02  0.02    5.36  1.30  1.12 -0.06 -0.08 -0.09  
5    0.10  0.14    7.74  1.48  1.41  0.03  0.04  0.06  
..    ...   ...     ...   ...   ...   ...   ...   ...  
128  0.12  0.17    7.56  2.07  1.45  0.06  0.13  0.19  
129  0.09  0.26    5.99  1.27  2.74  0.06  0.08  0.21  
130  0.12  0.26    7.14  1.89  2.10  0.07  0.12  0.26  
131  0.10  0.15  170.96  4.55  1.45  0.02  0.10  0.14  
132  0.07  0.10    5.44  1.84  1.40  0.05  0.08  0.12  

[132 rows x 26 columns]

In [4]:
X=bank.drop(['D','YR'],axis=1)	

In [5]:
X.shape

(132, 24)

In [6]:
y=bank['D'].values

In [7]:

scaler = MinMaxScaler()
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, 
                                                    random_state=23,stratify=y)
X_scl_trn = scaler.fit_transform(X_train) 
X_scl_tst = scaler.transform(X_test) 

In [8]:
X_torch = torch.from_numpy(X_scl_trn)
y_torch = torch.from_numpy(y_train)
print(X_torch.size())
print(y_torch.size())

torch.Size([92, 24])
torch.Size([92])


In [9]:
type(X_torch)

torch.Tensor

In [10]:
X_scl_trn.shape[1]

24

In [11]:
#torch.manual_seed(23)
# Create a model
#model = nn.Sequential(nn.Linear(in_features=X_scl_trn.shape[1], out_features=24),
 #                    nn.Linear(24, 15 ),
  #                    nn.ReLU(),
   #                   nn.Linear(15, 10),
    #                  nn.ReLU(),
     #                 nn.Linear(10,5),
      #                nn.ReLU(),
       #               nn.Linear(5,1),
        #              nn.Sigmoid())

In [12]:
#model.parameters

In [13]:
#criterion = torch.nn.BCELoss()
#optimizer = torch.optim.SGD(model.parameters(), lr = 0.001)
#optimizer

In [14]:
#y_torch.size()

In [15]:
#y_pred = model(X_torch.float())
#y_torch = y_torch.unsqueeze(1)
#print(y_torch.shape)
#print(y_pred.shape)

In [16]:
#y_pred[:5]

In [17]:
#loss = criterion(y_pred, y_torch.float())
#loss

In [18]:
#for epoch in np.arange(0,1000):
       # Forward pass: Compute predicted y by passing x to the model
 #      y_pred_prob = model(X_torch.float())

       # Compute and print loss
#     loss = criterion(y_pred_prob, y_torch.float())
 #      if epoch%100 == 0:
  #        print('epoch: ', epoch+1,' loss: ', loss.item())

       # Zero gradients, perform a backward pass, and update the weights.
   #    optimizer.zero_grad()

       # perform a backward pass (backpropagation)
    #   loss.backward()

       # Update the parameters
     #  optimizer.step()
#print('epoch: ', epoch+1,' loss: ', loss.item())

In [19]:
#loss

In [20]:
#X_torch_test = torch.from_numpy(X_scl_tst)
#type(X_torch_test)

In [21]:
#y_pred_prob = model(X_torch_test.float()) # Equivalent predict_proba / predict
#type(y_pred_prob)

In [22]:
#y_pred_prob

In [23]:
#y_pred_prob = nn.Sigmoid()(y_pred_prob)
#y_pred_prob

In [24]:
#y_pred_prob = y_pred_prob.detach().numpy()
#type(y_pred_prob)

In [25]:
#y_pred = np.where(y_pred_prob >= 0.5,1,0)

#y_pred[:5]

In [26]:
#print(accuracy_score(y_test,y_pred))

In [27]:
#log_loss(y_test, y_pred_prob)

0.7129465202106335 & accuracy_score=0.5

In [28]:
torch.manual_seed(23)
# Create a model
model = nn.Sequential(nn.Linear(in_features=X_scl_trn.shape[1], out_features=24),
                      nn.ReLU(),
                      nn.Linear(24,13),
                      nn.ReLU(),
                      nn.Linear(13, 7),
                      nn.ReLU(),
                      nn.Linear(7,1),
    
                      nn.Sigmoid())

In [29]:
criterion = torch.nn.BCELoss()
optimizer = torch.optim.SGD(model.parameters(), lr = 0.001)
optimizer

SGD (
Parameter Group 0
    dampening: 0
    differentiable: False
    foreach: None
    lr: 0.001
    maximize: False
    momentum: 0
    nesterov: False
    weight_decay: 0
)

In [30]:
y_pred = model(X_torch.float())
y_torch = y_torch.unsqueeze(1)
print(y_torch.shape)
print(y_pred.shape)

torch.Size([92, 1])
torch.Size([92, 1])


In [31]:
loss = criterion(y_pred, y_torch.float())
loss

tensor(0.6986, grad_fn=<BinaryCrossEntropyBackward0>)

In [32]:
for epoch in np.arange(0,1000):
       # Forward pass: Compute predicted y by passing x to the model
       y_pred_prob = model(X_torch.float())

       # Compute and print loss
       loss = criterion(y_pred_prob, y_torch.float())
       if epoch%100 == 0:
          print('epoch: ', epoch+1,' loss: ', loss.item())

       # Zero gradients, perform a backward pass, and update the weights.
       optimizer.zero_grad()

       # perform a backward pass (backpropagation)
       loss.backward()

       # Update the parameters
       optimizer.step()
#print('epoch: ', epoch+1,' loss: ', loss.item())

epoch:  1  loss:  0.6985804438591003
epoch:  101  loss:  0.6981284022331238
epoch:  201  loss:  0.6977104544639587
epoch:  301  loss:  0.6973186731338501
epoch:  401  loss:  0.6969509720802307
epoch:  501  loss:  0.6966062784194946
epoch:  601  loss:  0.6962832808494568
epoch:  701  loss:  0.6959800720214844
epoch:  801  loss:  0.6956954598426819
epoch:  901  loss:  0.6954275965690613


In [33]:
X_torch_test = torch.from_numpy(X_scl_tst)
type(X_torch_test)

torch.Tensor

In [34]:
y_pred_prob = model(X_torch_test.float()) # Equivalent predict_proba / predict
type(y_pred_prob)

torch.Tensor

In [35]:
y_pred_prob = nn.Sigmoid()(y_pred_prob)
y_pred_prob

tensor([[0.6134],
        [0.6137],
        [0.6136],
        [0.6137],
        [0.6136],
        [0.6139],
        [0.6135],
        [0.6135],
        [0.6135],
        [0.6138],
        [0.6139],
        [0.6139],
        [0.6138],
        [0.6137],
        [0.6138],
        [0.6135],
        [0.6137],
        [0.6138],
        [0.6136],
        [0.6137],
        [0.6134],
        [0.6135],
        [0.6139],
        [0.6137],
        [0.6137],
        [0.6140],
        [0.6137],
        [0.6137],
        [0.6138],
        [0.6137],
        [0.6135],
        [0.6136],
        [0.6135],
        [0.6139],
        [0.6135],
        [0.6134],
        [0.6135],
        [0.6137],
        [0.6139],
        [0.6137]], grad_fn=<SigmoidBackward0>)

In [36]:
y_pred_prob = y_pred_prob.detach().numpy()
type(y_pred_prob)

numpy.ndarray

In [37]:
y_pred = np.where(y_pred_prob >= 0.5,1,0)

y_pred[:5]

array([[1],
       [1],
       [1],
       [1],
       [1]])

In [38]:
print(accuracy_score(y_test,y_pred))

0.5


In [39]:
log_loss(y_test, y_pred_prob)

0.7196042529753102

accuracy_score(y_test,y_pred)-0.5<br>
log_loss(y_test, y_pred_prob)-0.7196042529753102

Consider the dataset  in Cases/Concrete Strength Y=Strength. Run the MLP with PyTorch Code with Layers as ( features, 5,3,2 , 1)   and  ( features, 10, 5, 1)

In [60]:
con=pd.read_csv("Concrete_Data.csv")

In [61]:
con

Cement  Blast    Fly  Water  Superplasticizer  Coarse   Fine  Age  \
0      540.0    0.0    0.0  162.0               2.5  1040.0  676.0   28   
1      540.0    0.0    0.0  162.0               2.5  1055.0  676.0   28   
2      332.5  142.5    0.0  228.0               0.0   932.0  594.0  270   
3      332.5  142.5    0.0  228.0               0.0   932.0  594.0  365   
4      198.6  132.4    0.0  192.0               0.0   978.4  825.5  360   
...      ...    ...    ...    ...               ...     ...    ...  ...   
1025   276.4  116.0   90.3  179.6               8.9   870.1  768.3   28   
1026   322.2    0.0  115.6  196.0              10.4   817.9  813.4   28   
1027   148.5  139.4  108.6  192.7               6.1   892.4  780.0   28   
1028   159.1  186.7    0.0  175.6              11.3   989.6  788.9   28   
1029   260.9  100.5   78.3  200.6               8.6   864.5  761.5   28   

      Strength  
0        79.99  
1        61.89  
2        40.27  
3        41.05  
4        44.30  
...        ...  
1025     44.28  
1026     31.18  
1027     23.70  
1028     32.77  
1029     32.40  

[1030 rows x 9 columns]

In [62]:
X=con.drop(['Strength'],axis=1)
y=con['Strength'].values

In [63]:

#scaler = MinMaxScaler()
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, 
                                                    random_state=23)
#X_scl_trn = scaler.fit_transform(X_train) 
#X_scl_tst = scaler.transform(X_test) 

In [64]:
X_torch = torch.from_numpy(X_scl_trn)
y_torch = torch.from_numpy(y_train)
print(X_torch.size())
print(y_torch.size())

torch.Size([721, 8])
torch.Size([721])


In [79]:
torch.manual_seed(23)
# Create a model
model = nn.Sequential(nn.Linear(in_features=X_scl_trn.shape[1], out_features=8),
                      nn.ReLU(),
                      nn.Linear(5,3),
                      nn.ReLU(),
                      nn.Linear(3, 2),
                      nn.ReLU(),
                      nn.Linear(2,1))

In [80]:
criterion = torch.nn.BCELoss()
optimizer = torch.optim.SGD(model.parameters(), lr = 0.001)
optimizer

SGD (
Parameter Group 0
    dampening: 0
    differentiable: False
    foreach: None
    lr: 0.001
    maximize: False
    momentum: 0
    nesterov: False
    weight_decay: 0
)

In [81]:
#y_pred_prob = model(X_torch_test.float()) # Equivalent predict_proba / predict
#type(y_pred_prob)

In [82]:
#y_pred_prob = nn.ReLU()(y_pred_prob)
#y_pred_prob

In [83]:
#y_pred_prob = y_pred_prob.detach().numpy()
#type(y_pred_prob)

In [84]:
y_pred = np.where(y_pred_prob >= 0.5,1,0)

y_pred[:5]

array([[1],
       [1],
       [1],
       [1],
       [1]])

In [85]:
print(r2_score(y_test,y_pred))

ValueError: Found input variables with inconsistent numbers of samples: [309, 40]